File: trunk/Database/nastya/src/database/search/DatabaseIterator.java
package database.search;

import tree.Key;
import tree.BPlusTree;
import tree.TreeLeaf;
import tree.IndexableData;
import dbentities.Condition;
import dbentities.Card;
import database.index.AddressData;
import database.index.DatabaseKey;

import java.util.Vector;

/**
 * Итератор по базе данных
 *
 * @author nastya
 *         Date: 30.08.2009
 *         Time: 3:9:17
 */
public class DatabaseIterator {
    // текущий ключ
    private DatabaseKey currentKey;
    //если при сравнении с этим условием возвращается положительное число, то
    // итератор прекращает проход
    private Condition stopCondition;
    private BPlusTree<Card, DatabaseKey,AddressData> tree;

    public DatabaseIterator(Condition startCondition, Condition stopCondition, BPlusTree<Card, DatabaseKey, AddressData> tree) {
        this.stopCondition = stopCondition;
        this.tree = tree;
        this.currentKey = tree.find(new DatabaseKey(startCondition));
    }

    /**
     * Есть ли еще элементы, которые может вернуть итератор
     * @return <code>true</code>, если в итеаторе еще  есть элементы; <code>false</code> иначе
     */
    public boolean hasNext(){
        return currentKey != null;
    }


    /**
     * Берет следующий ключ дерева на нижнем уровне, если ключи еще есть и
     * условие останова не дстигнуто и возвращает данные из листа, прикрепленного к нему
     * @return
     */
    public Vector<AddressData> next(){
        Key oldKey = currentKey;
        //Ошибка использования итератора
        if(oldKey == null || !(oldKey.getLink() instanceof TreeLeaf)){
            throw new IteratorException();
        }
        //Берем следующий ключ в дереве (в уровне)
        if(currentKey != null){
            currentKey = tree.keyAfter(currentKey);
            // если условие останова достигнуто, то итератор больше не возвращает ключей
            if(currentKey != null && currentKey.compareTo(new DatabaseKey(stopCondition)) > 0){
                currentKey = null;
            }
        }

        return ((TreeLeaf<AddressData>)oldKey.getLink()).getAllData();
    }


}

File: branches/2008-2009/Database/Babaka/DB/DB3/Iterator/DatabaseIterator.cs
using System;
using System.Collections;
using DB3.Entities;
using DB3.Index;
using DB3.Interfaces;
using DB3.Tree;
using DB3.Util;

namespace DB3.Iterator
{
    internal class DatabaseIterator : IEnumerable
    {
        private DatabaseKey _currentKey;

        private Condition _stopCondition;
        private BTree<Card, DatabaseKey, AddressData> _tree;

        private int _position;

        public DatabaseIterator(Condition startCondition, Condition stopCondition, BTree<Card, DatabaseKey, AddressData> tree)
        {
            _stopCondition = stopCondition;
            _tree = tree;
            _currentKey = tree.Find(new DatabaseKey(startCondition));
        }

        public bool HasNext()
        {
            return _currentKey != null;
        }

        public SortedList<AddressData> Next()
        {
            Key oldKey = _currentKey;
            if (oldKey == null || !(oldKey.Link is Leaf<AddressData>))
            {
                throw new IteratorException();
            }
            //Берем следующий ключ в дереве (в уровне)
            if (_currentKey != null)
            {
                _currentKey = _tree.KeyAfter(_currentKey);
                // если условие останова достигнуто, то итератор больше не возвращает ключей
                if (_currentKey != null && _currentKey.CompareTo(new DatabaseKey(_stopCondition)) > 0)
                {
                    _currentKey = null;
                }
            }

            return ((Leaf<AddressData>)oldKey.Link).GetAllData();
        }

        public IEnumerator GetEnumerator()
        {
            throw new NotImplementedException();
        }
    }
}
